File: DSA/ArrayList/ArrayList.php
<?php

namespace DSA\ArrayList;

use InvalidArgumentException;
use OutOfBoundsException;
use UnderflowException;

class ArrayList
{
    private array $data;

    private int $size;

    public function __construct(private int $capacity = 4)
    {
        $this->size = 0;
        $this->data = array_fill(0, $this->capacity, null);
    }

    public function size(): int
    {
        return $this->size;
    }

    public function capacity(): int
    {
        return $this->capacity;
    }

    public function isEmpty(): bool
    {
        return $this->size === 0;
    }

    public function get(int $index): mixed
    {
        $this->checkOutOfBounds($index);

        return $this->data[$index];
    }

    public function set(int $index, mixed $item): void
    {
        $this->checkOutOfBounds($index);

        $this->data[$index] = $item;
    }

    public function append($item): void
    {
        $this->checkCapacityReached();

        $this->data[$this->size++] = $item;
    }

    public function pop(): mixed
    {
        $this->checkIsEmpty();

        $item = $this->data[--$this->size];
        unset($this->data[$this->size]);

        return $item;
    }

    public function insert(int $index, mixed $item): void
    {
        if ($index > $this->size || $index < 0) {
            throw new OutOfBoundsException("Index '$index' out of bounds");
        }

        $this->checkCapacityReached();
        $this->size++;

        for ($i = $this->size - 2; $i >= $index; $i--) {
            $this->data[$i + 1] = $this->data[$i];
        }

        $this->set($index, $item);
    }

    public function prepend(mixed $item): void
    {
        $this->insert(0, $item);
    }

    public function delete(int $index): void
    {
        $this->checkOutOfBounds($index);

        for ($i = $index; $i < $this->size - 1; $i++) {
            $this->set($i, $this->get($i + 1));
        }

        unset($this->data[--$this->size]);
    }

    public function remove(mixed $item): void
    {
        $index = $this->find($item);
        if ($index === -1) {
            throw new InvalidArgumentException("Item '$item' not found");
        }

        $this->delete($index);
    }

    public function find(mixed $item): int
    {
        for ($i = 0; $i < $this->size; $i++) {
            if ($this->get($i) === $item) {
                return $i;
            }
        }

        return -1;
    }

    private function resize(): void
    {
        $this->capacity *= 2;
        $newArrayList = array_fill(0, $this->capacity, null);

        for ($i = 0; $i < $this->size; $i++) {
            $newArrayList[$i] = $this->data[$i];
        }

        $this->data = $newArrayList;
    }

    private function checkOutOfBounds(int $index): void
    {
        if ($index >= $this->size || $index < 0) {
            throw new OutOfBoundsException("Index '$index' out of bounds");
        }
    }

    private function checkCapacityReached(): void
    {
        if ($this->size === $this->capacity) {
            $this->resize();
        }
    }

    private function checkIsEmpty(): void
    {
        if ($this->isEmpty()) {
            throw new UnderflowException('ArrayList is empty');
        }
    }
}

File: DSA/java/src/main/java/ds/ArrayList.java
package ds;

public class ArrayList<T> {
    private Object[] data;
    private int size = 0;
    private int capacity = 4;

    public ArrayList() {
        this(4);
    }

    public ArrayList(int capacity) {
        this.capacity = capacity;
        this.data = new Object[capacity];
    }

    public int size() {
        return size;
    }

    public int capacity() {
        return capacity;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public T get(int index) {
        checkIndex(index);

        return data(index);
    }

    public void set(int index, T item) {
        checkIndex(index);

        data[index] = item;
    }

    public void append(T item) {
        ensureCapacity();

        data[size++] = item;
    }

    public T pop() {
        if (isEmpty()) {
            throw new IllegalStateException("Can't pop on an empty ArrayList");
        }

        T item = data(--size);
        data[size] = null;

        return item;
    }

    public void insert(int index, T item) {
        if (index > size || index < 0) {
            throw new IndexOutOfBoundsException("Index: " + index);
        }

        ensureCapacity();

        size++;

        for (int i = size - 2; i >= index; i--) {
            data[i + 1] = data[i];
        }

        set(index, item);
    }

    public void prepend(T item) {
        insert(0, item);
    }

    public void delete(int index) {
        checkIndex(index);

        for (int i = index; i < size - 1; i++) {
            data[i] = data[i + 1];
        }

        data[--size] = null;
    }

    public int find(T item) {
        for (int i = 0; i < size; i++) {
            if (data[i] == item) {
                return i;
            }

        }

        return -1;
    }

    public void remove(T item) {
        int index = find(item);
        if (index == -1) {
            throw new IllegalArgumentException("Item not found on the list");
        }

        delete(index);
    }

    private void checkIndex(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("Index: " + index);
        }
    }

    private void ensureCapacity() {
        if (size == capacity) {
            resize();
        }
    }

    private void resize() {
        capacity *= 2;
        Object[] newList = new Object[capacity];

        for (int i = 0; i < size; i++) {
            newList[i] = data[i];
        }

        data = newList;
    }

    @SuppressWarnings("unchecked")
    private T data(int index) {
        return (T) data[index];
    }
}
